src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,165 @@
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                s - index);
+        elementData[index] = value;
+        size = s + 1;
 
     }
 
     @Override
     public void addAll(List<T> list) {
 
+        Object[] a = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            a[i] = list.get(i);
+        }
+        int numNew = a.length;
+        if (numNew == 0) {
+            return;
+        }
+        Object[] elementData;
+        final int s;
+        if (numNew > (elementData = this.elementData).length - (s = size)) {
+            elementData = grow(s + numNew);
+        }
+        int numMoved = s - 1;
+        if (numMoved > 0) {
+            System.arraycopy(elementData, size - 1,
+                    elementData, numNew,
+                    numMoved);
+        }
+        System.arraycopy(a, 0, elementData, size, numNew);
+        size = s + numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForAdd(index + 1);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
+        rangeCheckForAdd(index + 1);
 
+        //  T oldValue = elementData[index];
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForAdd(index + 1);
+        final Object[] es = elementData;
+        @SuppressWarnings("unchecked") T oldValue = (T) es[index];
+        fastRemove(es, index);
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        Object result;
+        final int size = this.size;
+        int i = 0;
+        found:
+        {
+            if (element == null) {
+                for (; i < size; i++) {
+                    if (elementData[i] == null) {
+                        break found;
+                    }
+                }
+            } else {
+                for (; i < size; i++) {
+                    if (element.equals(elementData[i])) {
+                        break found;
+                    }
+                }
+            }
+            throw new NoSuchElementException();
+        }
+        result = elementData[i];
+        fastRemove(elementData, i);
+        return (T) result;
+    }
+
+    private void fastRemove(Object[] es, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(es, i + 1, es, i, newSize - i);
+        }
+        es[size = newSize] = null;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Object[] grow(int minCapacity) {
```

Comment:
No need to return new array because you already have access to it from all parts of this class. 
```suggestion
    private void grow() {
```